### PR TITLE
Introduce ErrorFormatter for GitHub Actions

### PR DESCRIPTION
**Pull request by Quetzacoalt91**
This PR allows PHPStan to report errors on the GitHub interface when used with GitHub Actions, in pull-requests changes. We started using this formatter at PrestaShop to help our contributor to find the required changes easily (https://github.com/PrestaShop/php-dev-tools/pull/24).

image

Note the only displayed errors are targeting a specific line of a file, generic errors (= Not file specific) are not concerned. For developers looking at the complete report in the log, the class `TableErrorFormatter` is used so they can read the default view.

### CODE REVIEW

File: src/Command/ErrorFormatter/GithubErrorFormatter.php [hunk 1]
```diff
@@ -0,0 +1,49 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Command\ErrorFormatter;
+
+use PHPStan\Command\AnalysisResult;
+use PHPStan\Command\Output;
+use PHPStan\File\RelativePathHelper;
+
+class GithubErrorFormatter extends TableErrorFormatter
```
Comment (ondrejmirtes) [inline src/Command/ErrorFormatter/GithubErrorFormatter.php:9]: Composition over inheritance should be used. Ask for TableErrorFormatter in the constructor, do not extend from it.

File: src/Command/ErrorFormatter/GithubErrorFormatter.php [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+			$metas = [
+				'file' => $this->relativePathHelper->getRelativePath($fileSpecificError->getFile()),
+				'line' => $fileSpecificError->getLine(),
+				'col' => 0,
+			];
+			array_walk($metas, static function (&$value, string $key): void {
+				$value = sprintf('%s=%s', $key, (string) $value);
+			});
+
+			$severity = $fileSpecificError->canBeIgnored() ? 'warning' : 'error';
```
Comment (ondrejmirtes) [inline src/Command/ErrorFormatter/GithubErrorFormatter.php:37]: All should be errors.

**Comment by Quetzacoalt91** [comment]
While I'm applying the changes, you can find more details about the annotations here: https://help.github.com/en/actions/reference/workflow-commands-for-github-actions#setting-an-error-message

**Comment by ondrejmirtes** [comment]
Would be a good idea to link it in the implementation :)

**Comment by Quetzacoalt91** [comment]
Changes have been applied. About:

> What happens about errors reported on files that haven't changed in a pull request?

GitHub will display them anyway in a dedicated space, below the diff. Unfortunately I don't have any link to provide as an example (the PR was updated), only a screenshot I took while PHPStan was in action.

image

**Comment by ondrejmirtes** [comment]
That’s really nice! Could we somehow incorporate notFileSpecificErrors too? How GitHub behaves when we add an error to a nonexistent file?

**Comment by Quetzacoalt91** [comment]
We could add them in the report, but I didn't in the original version as finding them is not an intuitive task. These issues are reported in the action tab when we select the related workflow: 

image

**Comment by staabm** [comment]
To achieve the same effect I usually use the checkstyle output format and https://github.com/staabm/annotate-pull-request-from-checkstyle

**Comment by ondrejmirtes** [comment]
@Quetzacoalt91 I really like how it looks! Besides notFileSpecificErrors, we could also present warnings as actual warnings. You can create a warning for example by not escaping regex-special characters in `ignoreErrors`:

* Generates a warning: `#Result of || is always false#`
* Does not generate a warning: `#Result of \|\| is always false#`

@staabm I'm aware of this but built-in formatter creates new opportunities like do this by default if we detect we're running in GitHub Actions.

**Comment by Quetzacoalt91** [comment]
All right, at the same time I added the warning you were talking about, I added the generic errors as GitHub handle them in a way.

**Comment by ondrejmirtes** [comment]
Thank you, I love it. Will merge it if the build passes. I'll also try to make it the default, but I want to tackle that myself.

**Comment by Quetzacoalt91** [comment]
Thanks for your time with the improvements, I’m glad you like it. :)

**Comment by ondrejmirtes** [comment]
FYI https://github.com/phpstan/phpstan-src/commit/6e7d5f6ef904f27863db99ee823e58ccbb689e7f + https://github.com/phpstan/phpstan-src/commit/6fd85e3b7dae790061b278852b1b34a17e3679c6

**Comment by b1rdex** [comment]
Is the same or similar possible for Gitlab?

**Comment by ondrejmirtes** [comment]
Doesn't look like so: https://twitter.com/johannpardanaud/status/1278273861088395264

There's already a `gitlab` error formatter but I think it's limited to the paid version.

**Pull request merged**